### ISSUE TITLE
What causes mapHS2FCL not to run on parallel processes?

### ISSUE THREAD

**Issue by chrmongeau** [SWS-Methodology/faoswsTrade#116]
There *seems* to be two steps in the HS->FCL mapping: one of them runs in parallel, while the other doesn't (and it's slower). I say this because I see the progress bar in the second step and if there is a progress bar it means that the code is not running in parallel processes.

@malexan: do you have any hints?

**Comment by malexan** [comment]
Does this second part takes too long? I mean like the first part? 

There are several steps in `mapHS2FCL` and only two of them use parallel backend.

**Comment by malexan** [comment]
By the way, does now SWS itself use parallel computation?

**Comment by chrmongeau** [comment]
> Does this second part takes too long? I mean like the first part?

No, it takes a lot more. Something like 40 minutes.

> By the way, does now SWS itself use parallel computation?

AFAIK yes, but @sebastian-c can confirm.

**Comment by malexan** [comment]
40 minutes? 8-0 And how much it takes to run the module for the whole year? Is it on Windows with 8 cores?

We could create a separate logger and send there all technical details (not to mix with methodology stuff).

**Comment by malexan** [comment]
On my 4 cores of notebook CPU full trade module run takes about 20 minutes.

Of course it is without real SWS reading/writing.

**Comment by malexan** [comment]
There is a piece in `sel1FCL()` for vectorization. It uses `do()` (which can show that progress bar). Maybe this `do()` is not too efficient. 

```r
  if(length(unique(hsfclmatch$datumid)) > 1L) return(
    hsfclmatch %>%
      group_by_(~datumid) %>%
      do(sel1FCL(., maptable)) %>%
      ungroup
  )
```

**Comment by sebastian-c** [comment]
I do not believe that it does use parallel processing, but I guess you could do it. On QA, we have:
```
QA:
model name	: Intel(R) Xeon(R) CPU           X5690  @ 3.47GHz

Production:
model name	: Intel(R) Xeon(R) CPU E5-2690 v2 @ 3.00GHz
```
As a side note, please make sure you disable progress bars on the system (You can use `CheckDebug` to check), they dump every update of the progress bar to the logs.

**Comment by malexan** [comment]
Now we have such passage in `main.R`:
```r

# Register CPU cores ####
if(multicore) {
  if(all(c("doParallel", "foreach") %in%
         rownames(installed.packages()))) {

    flog.debug("Multicore backend is available.")

    cpucores <- parallel::detectCores(all.tests = TRUE)

    flog.debug("CPU cores detected: %s.", cpucores)

    doParallel::registerDoParallel(cores = cpucores)
  } else {
    flog.debug("Multicore backend is not available.")

    multicore <- FALSE
  }
}
```

And `multicore <- TRUE` in the beginning.

As about progress bars, there is a strange behavior of `dplyr::do()` even `dplyr.show_progress <- FALSE`. Working on this.

**Comment by chrmongeau** [comment]
> By the way, does now SWS itself use parallel computation?

This is on the server:

```
> parallel::detectCores()
[1] 2
```

I was told that a request for improving this will be done. I don't know how much time it will require, but I suppose it won't be quick.

**Comment by sebastian-c** [comment]
I'll double check with Giorgio as I'm fairly sure this issue has been raised before

**Comment by chrmongeau** [comment]
@malexan I ran the module again (2011) and the `tldatalinks` creation took around 50 minutes. I'll try to profile it, but I bet that the bottleneck is the `do()` in `sel1FCL()` (as you suggested).

**Comment by chrmongeau** [comment]
@malexan I just ran the module with `year = 2013` and the `tldataliks` creation was relatively fast: it took around 5/10 minutes to complete. The problem reported above may be related to 2011 for some reason.

**Comment by chrmongeau** [comment]
@malexan It took 1 hour and 30 minutes for 2008.

`dplyr::do()`'s help says that "If you're familiar with plyr, do with named arguments is basically equivalent to dlply, and do with a single unnamed argument is basically equivalent to ldply." ([source](http://dplyr.tidyverse.org/reference/do.html#connection-to-plyr)), so I think that `do()` could be replaced by one of  `dlply` or `ldply`. Unfortunately, I don't understand the function, so I won't attempt to improve it.

> On my 4 cores of notebook CPU full trade module run takes about 20 minutes.

I did a run on SWS (year = 2008) and the full trade module took 3 hours.

> As about progress bars, there is a strange behavior of `dplyr::do() even dplyr.show_progress <- FALSE`

Fixed in 3012c46

**Comment by chrmongeau** [comment]
> I did a run on SWS (year = 2008) and the full trade module took 3 hours.

I forgot:

`DEBUG [2017-04-13 18:46:59] Multicore backend is not available.`

**Comment by chrmongeau** [comment]
So, 3 hours for the full module on SWS without multicore, and 1 hour and a half on my local machine for just the HS>FCL mapping (again, I'm referring to 2008)... Probably something is wrong in my machine (16GB, 8 cores).

**Comment by malexan** [comment]
> DEBUG [2017-04-13 18:46:59] Multicore backend is not available.

That is important, thanks!

```r
  if(all(c("doParallel", "foreach") %in%
         rownames(installed.packages(noCache = TRUE)))) {
    flog.info("Multicore backend is available.", name = "dev")
...
  } else {
    flog.debug("Multicore backend is not available.", name = "dev")
    multicore <- FALSE
```

Does R in Windows use other packages for parallel processing than `doParallel and `foreach`?

**Comment by chrmongeau** [comment]
> Does R in Windows use other packages for parallel processing than doParallel andforeach`?

If you are asking because I reported that "Multicore backend is not available", it would not be important because that message was produced on SWS (i.e., on SWS we cannot use multicore, while I can do it on my machine). If you are asking because you think that there could be some issues with those packages on Windows, the answer is "I don't know". I can investigate, though. Anyway, I am using some other code that uses `parallel` and `doParallel` on Windows and they seem to work fine.

**Comment by malexan** [comment]
My message "Multicore backend is not available" is not correct. It should be changed to "doParallel and  foreach packages are not installed". So my current logic is that if these packages are not available than parallel execution setting is switched off. It works on Linux but maybe it is wrong for Windows (you say about parallel, but not about foreach). And we should introduce different checks for Linux and Windows.

It looks like you are to get `FALSE` on the following, right?

```r
all(c("doParallel", "foreach") %in% 
    rownames(installed.packages(noCache = TRUE))))
```

**Comment by malexan** [comment]
Can you give a short example of working parallel code on Windows?

**Comment by chrmongeau** [comment]
Sure. Here it goes:

```r
myfunction <- function(var) {
  Sys.sleep(1)
  return(var*2)
}

# Leave one core available
n_cores <- parallel::detectCores() - 1
cl <- parallel::makeCluster(n_cores)
doParallel::registerDoParallel(cl)

# If a library is needed
parallel::clusterEvalQ(cl, library(faosws))

# If variables/functions are needed
parallel::clusterExport(cl, 'myfunction')

multicore <- TRUE

myresults <- plyr::mdply(
                       data.frame(var = 1:100),
                       .fun = myfunction,
                       .parallel = multicore,
                       .progress = 'text'
                       )

parallel::stopCluster(cl)
```

**Comment by malexan** [comment]
Please try this:
```r
cpucores <- parallel::detectCores(all.tests = TRUE)
doParallel::registerDoParallel(cores = cpucores)
myresults <- plyr::ldply(
  1:100,
  .fun = function(var) var * 2,
  .parallel = TRUE
)

cpucores

"foreach" %in% row.names(installed.packages(noCache = TRUE))
```

**Comment by chrmongeau** [comment]
Got this:

```
> cpucores <- parallel::detectCores(all.tests = TRUE)
> doParallel::registerDoParallel(cores = cpucores)
> myresults <- plyr::ldply(
+   1:100,
+   .fun = function(var) var * 2,
+   .parallel = TRUE
+ )
Warning messages:
1: <anonymous>: ... may be used in an incorrect context: ‘.fun(piece, ...)’
 
2: <anonymous>: ... may be used in an incorrect context: ‘.fun(piece, ...)’
 
> 
> cpucores
[1] 8
> 
> "foreach" %in% row.names(installed.packages(noCache = TRUE))
[1] TRUE
```

**Comment by malexan** [comment]
As about SWS, are "foreach" and "doParallel" installed there?

**Comment by chrmongeau** [comment]
On SWS:

```
> c("doParallel", "foreach", "parallel") %in% rownames(installed.packages(noCache = TRUE))
[1] FALSE FALSE  TRUE
```

**Comment by malexan** [comment]
OK, it explains why parallel execution is not available on SWS. @sebastian-c, is it possible to install `doParallel` and `foreach`?

**Comment by sebastian-c** [comment]
Anything's possible - but foreach doesn't work on Windows as far as I know (it requires forking which Windows doesn't support) but doParallel seems to according to [its vignette](https://cran.r-project.org/web/packages/doParallel/vignettes/gettingstartedParallel.pdf). You should be able to install them into your home directory if you want to test them.

**Comment by malexan** [comment]
But `foreach` package is available for Windows and it is used by `plyr` package:

> .paropts	
> a list of additional options passed into the foreach function when parallel computation is enabled. 

And this is from `doParallel` vignette:

> The doParallel package acts as an interface between foreach and the parallel package of R
2.14.0 and later. The parallel package is essentially a merger of the multicore package, which
was written by Simon Urbanek, and the snow package, which was written by Luke Tierney and
others. The multicore functionality supports multiple workers only on those operating systems
that support the fork system call; this excludes Windows. By default, doParallel uses multicore
functionality on Unix-like systems and snow functionality on Windows.

So it looks like `foreach` can work on Windows, but together with `snow` package.

Really the final purpose is to make a cross-platform piece of code, what would allow `plyr`'s functions to work in parallel (i.e. based on quotes above to work with `foreach`).
